sstables/kl/reader_impl.hh

```diff
@@ -170,277 +144,248 @@ private:
     uint32_t _ttl, _expiration;
 
     bool _shadowable;
+
+    using semaphore_type = seastar::basic_semaphore<semaphore_default_exception_factory>;
+    semaphore_type _production_sem;
+    future<> _production;
+    bool* _production_end;
 public:
     using consumer = row_consumer;
-    bool non_consuming() const {
-        return (((_state == state::DELETION_TIME_3)
-                || (_state == state::CELL_VALUE_BYTES_2)
-                || (_state == state::ATOM_START_2)
-                || (_state == state::ATOM_MASK_2)
-                || (_state == state::STOP_THEN_ATOM_START)
-                || (_state == state::COUNTER_CELL_2)
-                || (_state == state::RANGE_TOMBSTONE_4)
-                || (_state == state::EXPIRING_CELL_3)));
-    }
 
-    // process() feeds the given data into the state machine.
-    // The consumer may request at any point (e.g., after reading a whole
-    // row) to stop the processing, in which case we trim the buffer to
-    // leave only the unprocessed part. The caller must handle calling
-    // process() again, and/or refilling the buffer, as needed.
-    data_consumer::processing_result process_state(temporary_buffer<char>& data) {
-        try {
-            return do_process_state(data);
-        } catch (malformed_sstable_exception& exp) {
-            throw malformed_sstable_exception(exp.what(), _sst->get_filename());
+private:
+    future<> produce_64() {
+        co_await read_nonempty();
+        data_consumer::read_status rs = read_64(_data);
+        while (rs != data_consumer::read_status::ready) {
+            assert(_data.size() == 0);
+            co_await read_nonempty();
+            rs = primitive_consumer::consume(_data);
```

Comment:
I was afraid of this. A coroutine always allocates, so each produce_64() is an allocation.

In the existing code, reading some primitive has some overhead even in the common case where the data is ready in the buffer, but it's just some checks. Now we have to allocate coroutine state each time.

Did you measure performance? Not sure what would make a good test for this.

Reply:
Perhaps the primitives can be converted to continuations. But we'll still see overhead from futures, which are too expensive for such simple calculations.

Reply:
> 
> Did you measure performance? Not sure what would make a good test for this.

perhaps test/perf/perf_sstable after some tweaking

Reply:
I think if coroutines can be used here, they would be custom coroutines, not seastar coroutines. We'd replace proceed::no with `co_yield proceed::no`. The coroutine return value would be a custom wrapper around the handle.

It's still problematic. The entire parser would have to be one giant function (so we have one allocation per parse).

Reply:
(it would replace the big switch and gotos with standard control flow, but that's about it)

Reply:
For testing you can use perf_fast_forward where it scans multi partition data set.

@avikivity Why do you want to use co_yield with proceed::no?

Reply:
I want to replace the switch/goto based control flow with nested loops for clarity. But I'm not willing to give up performance for it, and the switch thing is pretty much optimal.

This is likely quite a hard task.